File: prac_final_1/mini_hoja_calculo.py
from collections import deque
from copy import deepcopy

# Estructuras de datos globales
hoja = []
indice_hash = {}
historial = []
cola_atencion = deque()

def calcular_promedio(n1, n2, n3):
    return (n1 + n2 + n3) / 3.0

def actualizar_hash():
    indice_hash.clear()

    for idx, estudiante in enumerate(hoja):
        indice_hash[estudiante["codigo"]] = idx

def guardar_historial():
    # Uso de deepcopy para aislar estados mutables dentro del Stack
    historial.append(deepcopy(hoja))

def registrar_estudiante():
    print("--- REGISTRAR NUEVO ESTUDIANTE ---")

    codigo = input("Codigo unico: ").strip()

    if codigo in indice_hash:
        print("[ERROR] Llave primaria duplicada. El codigo ya existe.")
        return

    nombre = input("Apellidos y Nombres: ").strip()

    try:
        nota1 = float(input("Nota 1 (0-20): "))
        nota2 = float(input("Nota 2 (0-20): "))
        nota3 = float(input("Nota 3 (0-20): "))
    except ValueError:
        print("[ERROR] Valores numericos incorrectos en calificaciones.")
        return

    estudiante = {
        "codigo": codigo,
        "nombre": nombre,
        "nota1": nota1,
        "nota2": nota2,
        "nota3": nota3,
        "promedio": round(calcular_promedio(nota1, nota2, nota3), 2)
    }

    guardar_historial()
    hoja.append(estudiante)

    actualizar_hash()

    print("[EXITO] Estudiante registrado e indexado en la estructura asociativa.")

def mostrar_hoja():
    if not hoja:
        print("[AVISO] La hoja de calculo no contiene registros activos.")
        return
    print(" " + "=" * 70)
    print(f"{'MINI HOJA DE CALCULO ACADEMICA':^70}")
    print("=" * 70)

    print(
        f"{'Codigo':<12}"
        f"{'Nombre Completo':<25}"
        f"{'N1':<8}"
        f"{'N2':<8}"
        f"{'N3':<8}"
        f"{'Promedio':<10}"
    )

    print("-" * 70)

    for e in hoja:
        print(
            f"{e['codigo']:<12}"
            f"{e['nombre']:<25}"
            f"{e['nota1']:<8.1f}"
            f"{e['nota2']:<8.1f}"
            f"{e['nota3']:<8.1f}"
            f"{e['promedio']:<10.2f}"
        )

    print("=" * 70)

def buscar_estudiante():
    codigo = input(
        "Ingrese codigo a buscar en la estructura asociativa: "
    ).strip()

    if codigo in indice_hash:
        pos = indice_hash[codigo]
        e = hoja[pos]

        print(
            f"[ENCONTRADO O(1)] "
            f"{e['nombre']} | "
            f"Promedio Evaluado: {e['promedio']:.2f}"
        )
    else:
        print("[AVISO] El codigo buscado no existe en la base indexada.")

def quicksort(lista):
    if len(lista) <= 1:
        return lista

    pivote = lista[-1]

    # Clasificación de Mayor a Menor (Descendente por Promedio)
    mayores = [
        x for x in lista[:-1]
        if x["promedio"] >= pivote["promedio"]
    ]

    menores = [
        x for x in lista[:-1]
        if x["promedio"] < pivote["promedio"]
    ]

    return quicksort(mayores) + [pivote] + quicksort(menores)

def merge_sort(lista):
    if len(lista) <= 1:
        return lista

    medio = len(lista) // 2

    izquierda = merge_sort(lista[:medio])
    derecha = merge_sort(lista[medio:])

    return merge(izquierda, derecha)

def merge(izquierda, derecha):
    resultado = []
    i = j = 0

    # Clasificación de Menor a Mayor alfanumérico
    # (Ascendente por Código)
    while i < len(izquierda) and j < len(derecha):
        if izquierda[i]["codigo"] <= derecha[j]["codigo"]:
            resultado.append(izquierda[i])
            i += 1
        else:
            resultado.append(derecha[j])
            j += 1

    resultado.extend(izquierda[i:])
    resultado.extend(derecha[j:])

    return resultado

def ordenar_por_promedio():
    global hoja

    if not hoja:
        return

    guardar_historial()

    hoja = quicksort(hoja)

    actualizar_hash()
    print("[OK] Matriz ordenada por Promedio via QuickSort Funcional.")

def ordenar_por_codigo():
    global hoja

    if not hoja:
        return

    guardar_historial()

    hoja = merge_sort(hoja)

    actualizar_hash()

    print("[OK] Matriz ordenada por Codigo via MergeSort Estable.")

def deshacer():
    global hoja

    if historial:
        hoja = historial.pop()

        actualizar_hash()

        print("[REVERSION] Estado de memoria restaurado a la version previa.")
    else:
        print("[INFO] El historial de modificaciones esta vacio.")

def agregar_cola_atencion():
    codigo = input(
        "Ingrese codigo del estudiante en espera: "
    ).strip()

    if codigo in indice_hash:
        cola_atencion.append(codigo)

        print(
            "[COLA] Estudiante insertado en la cola FIFO institucional."
        )
    else:
        print(
            "[ERROR] Codigo no registrado en las celdas vigentes."
        )

def atender_estudiante():
    if cola_atencion:
        codigo = cola_atencion.popleft()

        pos = indice_hash[codigo]

        print(
            f"[ATENCION] Despachando requerimiento de: "
            f"{hoja[pos]['nombre']}"
        )
    else:
        print(
            "[INFO] No se registran estudiantes pendientes en la cola."
        )

def suma_promedios_recursiva(i):
    if i == len(hoja):
        return 0.0

    return hoja[i]["promedio"] + suma_promedios_recursiva(i + 1)

def estadisticas():
    if not hoja:
        print("[AVISO] Datos insuficientes para el analisis.")
        return

    suma = suma_promedios_recursiva(0)
    promedio_general = suma / len(hoja)

    print(" " + "=" * 40)
    print(f" Estudiantes Registrados : {len(hoja)}")
    print(f" Promedio General Cohorte: {promedio_general:.2f}")
    print("=" * 40)

def menu():
    while True:
        print(" " + "=" * 45)
        print("    MENU PRINCIPAL - PYTHON CORE SYSTEMS    ")
        print("=" * 45)

        print("1. Registrar estudiante (Fila)")
        print("2. Mostrar hoja de calculo")
        print("3. Buscar estudiante por codigo (Hash)")
        print("4. Ordenar por promedio (QuickSort)")
        print("5. Ordenar por codigo (MergeSort)")
        print("6. Deshacer ultima accion (Stack)")
        print("7. Agregar a cola de espera (Queue)")
        print("8. Atender siguiente estudiante (FIFO)")
        print("9. Analitica y Estadisticas generales")
        print("10. Salir del programa")
        print("-" * 45)

        opcion = input("Seleccione una opcion (1-10): ").strip()

        if opcion == "1":
            registrar_estudiante()

        elif opcion == "2":
            mostrar_hoja()

        elif opcion == "3":
            buscar_estudiante()

        elif opcion == "4":
            ordenar_por_promedio()

        elif opcion == "5":
            ordenar_por_codigo()

        elif opcion == "6":
            deshacer()

        elif opcion == "7":
            agregar_cola_atencion()

        elif opcion == "8":
            atender_estudiante()

        elif opcion == "9":
            estadisticas()

        elif opcion == "10":
            print("Finalizando subprocesos interpretados... Salida.")
            break

        else:
            print("[ERROR] Codigo numerico no asignado en el menu.")

if __name__ == "__main__":
    menu()
        

File: prac_final_1/mini_hoja_calculo.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <stack>
#include <queue>
#include <iomanip>
#include <string>
#include <utility>

using namespace std;

struct Estudiante {
    string codigo;
    string nombre;
    float nota1;
    float nota2;
    float nota3;
    float promedio;
};

// Estructuras de almacenamiento global
vector<Estudiante> hoja;
unordered_map<string, int> indiceHash;
stack<vector<Estudiante>> historial;
queue<string> colaAtencion;

// Función en línea para cálculo aritmético eficiente
inline float calcularPromedio(float n1, float n2, float n3) {
    return (n1 + n2 + n3) / 3.0f;
}

// Reconstrucción del índice asociativo O(n)
void actualizarHash() {
    indiceHash.clear();

    for (size_t i = 0; i < hoja.size(); ++i) {
        indiceHash[hoja[i].codigo] = static_cast<int>(i);
    }
}

// Resguardo del estado actual previo a mutaciones
void guardarHistorial() {
    historial.push(hoja);
}

void registrarEstudiante() {
    Estudiante e;

    cout << "\n--- REGISTRAR NUEVO ESTUDIANTE --"<<endl;
    cout << "Codigo unico: "<<endl;
    cin >> e.codigo;

    // Validación de clave primaria duplicada en tiempo O(1)
    if (indiceHash.find(e.codigo) != indiceHash.end()) {
        cout << "\n[ERROR] El codigo ingresado ya se encuentra registrado."<<endl;
        return;
    }
        cin.ignore();

    cout << "\nApellidos y Nombres: "<<endl;
    getline(cin, e.nombre);

    cout << "\nNota 1: "<<endl;
    cin >> e.nota1;

    cout << "\nNota 2: "<<endl;
    cin >> e.nota2;

    cout << "\nNota 3: "<<endl;
    cin >> e.nota3;

    e.promedio = calcularPromedio(e.nota1, e.nota2, e.nota3);

    guardarHistorial();
    hoja.push_back(e);
    actualizarHash();

    cout << "\n[EXITO] Registro insertado e indexado correctamente."<<endl;
}

void mostrarHoja() {
    if (hoja.empty()) {
        cout << "\n[AVISO] La hoja de calculo no contiene registros."<<endl;
        return;
    }

    cout << "\n====================================================================";
    cout << "\n                   MINI HOJA DE CALCULO ACADEMICA                   ";
    cout << "\n====================================================================";

    cout << left << setw(12) << "\nCodigo"
         << setw(25) << "Nombre Completo"
         << setw(8)  << "N1"
         << setw(8)  << "N2"
         << setw(8)  << "N3"
         << setw(10) << "Promedio"
         << endl;

    cout << "-------------------------------------------------------------------"<<endl;

    for (const auto &e : hoja) {
        cout << left << setw(12) << e.codigo
             << setw(25) << e.nombre
             << setw(8)  << setprecision(1) << fixed << e.nota1
             << setw(8)  << e.nota2
             << setw(8)  << e.nota3
             << setw(10) << setprecision(2) << e.promedio<< endl;
    }

    cout << "\n===================================================================="<<endl;
}

void buscarEstudiante() {
    string codigo;

    cout << "\nIngrese el codigo a buscar en Tabla Hash: "<<endl;
    cin >> codigo;

    auto it = indiceHash.find(codigo);

    if (it != indiceHash.end()) {
        int pos = it->second;
        const auto &e = hoja[pos];

        cout << "\n[REGISTRO ENCONTRADO - BUSQUEDA O(1)] "<<endl;
        cout << "\nEstudiante: "
             << e.nombre
             << " | Promedio Actual: "
             << fixed 
             <<setprecision(2)
             << e.promedio
             << endl;
    } else {
        cout << "\n[AVISO] Codigo no registrado en la base indexada."<<endl;
    }
}

// QuickSort: Particionamiento por Promedio (Descendente de Mayor a Menor)
int particion(vector<Estudiante> &arr, int bajo, int alto) {
    float pivote = arr[alto].promedio;
    int i = bajo - 1;

    for (int j = bajo; j < alto; j++) {
        if (arr[j].promedio >= pivote) {
            i++;
            swap(arr[i], arr[j]);
        }
    }

    swap(arr[i + 1], arr[alto]);
    return i + 1;
}

void quickSort(vector<Estudiante> &arr, int bajo, int alto) {
    if (bajo < alto) {
        int pi = particion(arr, bajo, alto);

        quickSort(arr, bajo, pi - 1);   
        quickSort(arr, pi + 1, alto);
    }
}

// MergeSort: Combinación por Código Alfanumérico (Ascendente)
void merge(vector<Estudiante> &arr, int izq, int medio, int der) {
    int n1 = medio - izq + 1;
    int n2 = der - medio;

    vector<Estudiante> L(n1), R(n2);

    for (int i = 0; i < n1; i++)
        L[i] = arr[izq + i];

    for (int j = 0; j < n2; j++)
        R[j] = arr[medio + 1 + j];

    int i = 0, j = 0, k = izq;

    while (i < n1 && j < n2) {
        if (L[i].codigo <= R[j].codigo) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort(vector<Estudiante> &arr, int izq, int der) {
    if (izq < der) {
        int medio = izq + (der - izq) / 2;

        mergeSort(arr, izq, medio);
        mergeSort(arr, medio + 1, der);
        merge(arr, izq, medio, der);
    }
}

void ordenarPorPromedio() {
    if (hoja.empty())
        return;

    guardarHistorial();
        quickSort(hoja, 0, static_cast<int>(hoja.size() - 1));
    actualizarHash();

    cout << "[OK] Filas ordenadas por Promedio (Descendente) via QuickSort."<<endl;
}

void ordenarPorCodigo() {
    if (hoja.empty())
        return;

    guardarHistorial();

    mergeSort(hoja, 0, static_cast<int>(hoja.size() - 1));
    actualizarHash();

    cout << "[OK] Filas ordenadas por Codigo (Ascendente) via MergeSort."<<endl;
}

void deshacer() {
    if (!historial.empty()) {
        hoja = historial.top();
        historial.pop();

        actualizarHash();

        cout << "[UNDO] Estado de celdas restaurado al punto de control anterior."<<endl;
    } else {
        cout << "[INFO] No existen acciones registradas para deshacer."<<endl;
    }
}

void agregarColaAtencion() {
    string codigo;

    cout << "Ingrese el codigo del estudiante que entra a la cola: "<<endl;
    cin >> codigo;

    if (indiceHash.find(codigo) != indiceHash.end()) {
        colaAtencion.push(codigo);

        cout << "[COLA] Estudiante ingresado en la cola FIFO de secretaria."<<endl;
    } else {
        cout << "[ERROR] El codigo no figura registrado en la hoja actual."<<endl;
    }
}

void atenderEstudiante() {
    if (!colaAtencion.empty()) {
        string codigo = colaAtencion.front();
        colaAtencion.pop();

        int pos = indiceHash[codigo];

        cout << "[ATENCION] Despachando requerimiento de: "
             << hoja[pos].nombre
             << endl;
    } else {
        cout << "[INFO] Canal de atencion despejado. Cola vacia."<<endl;
    }
}

// Funcion Recursiva Pura para la acumulacion de promedios O(n)
float sumaPromediosRecursiva(size_t i) {
    if (i == hoja.size())
        return 0.0f;

    return hoja[i].promedio + sumaPromediosRecursiva(i + 1);
}

void estadisticas() {
    if (hoja.empty()) {
        cout << "[AVISO] Hoja sin datos para el calculo estadistico."<<endl;
        return;
    }
        float suma = sumaPromediosRecursiva(0);
    float promedioGeneral = suma / static_cast<float>(hoja.size());

    cout << "========= METRICAS ANALITICAS GENERALES ========="<<endl;
    cout << " Estudiantes Procesados : " << hoja.size()<< endl;
    cout << " Promedio General Coho. : "
         << fixed
         << setprecision(2)
         << promedioGeneral<< endl;
    cout << "================================================="<<endl;
}

void menu() {
    int opcion;

    do {
        cout << "============================================="<< endl;
        cout << "         MENU PRINCIPAL - SPREADSHEET        "<< endl;
        cout << "============================================="<< endl;
        cout << "1. Registrar Estudiante (Fila) "<< endl;
        cout << "2. Mostrar Hoja de Calculo "<< endl;
        cout << "3. Buscar Estudiante por Codigo (Hash) "<< endl;
        cout << "4. Ordenar por Promedio [QuickSort Desc] "<< endl;
        cout << "5. Ordenar por Codigo   [MergeSort Asc] "<< endl;
        cout << "6. Deshacer Ultima Modificacion [Stack] "<< endl;
        cout << "7. Insertar Estudiante a Cola de Atencion "<< endl;
        cout << "8. Atender Siguiente Estudiante [FIFO] "<< endl;
        cout << "9. Mostrar Estadisticas Consolidadas "<< endl;
        cout << "10. Salir de la Aplicacion "<< endl;
        cout << "--------------------------------------------"<< endl;
        cout << "Seleccione una opcion (1-10): "<< endl;

        if (!(cin >> opcion)) {
            cout << "[ALERTA] Entrada invalida.";

            cin.clear();
            cin.ignore(10000, ' ');

            continue;
        }

        switch (opcion) {
            case 1:
                registrarEstudiante();
                break;

            case 2:
                mostrarHoja();
                break;

            case 3:
                buscarEstudiante();
                break;

            case 4:
                ordenarPorPromedio();
                break;

            case 5:
                ordenarPorCodigo();
                break;

            case 6:
                deshacer();
                break;

            case 7:
                agregarColaAtencion();
                break;

            case 8:
                atenderEstudiante();
                break;

            case 9:
                estadisticas();
                break;
                            case 10:
                cout << "Liberando memoria interna... Programa cerrado.";
                break;

            default:
                cout << "[ERROR] Opcion fuera de rango.";
        }

    } while (opcion != 10);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    menu();

    return 0;
}